### PR TITLE
Refactor getClosestNeighbour

### PR DESCRIPTION
**Pull request by jtaveau**
I took the initiative to refactor this function as it's the most expensive one of the project (computationnally speaking). 
The results looked great on hotspot, but I had only a small bag of 15s to test it, would you have a bigger one? 

Before refactor:
Screenshot from 2024-02-11 13-26-14

After refactor:
Screenshot from 2024-02-11 13-08-32

### CODE REVIEW

**Comment by nachovizzo** [comment]
@jtaveau thanks a lot for the initiative. Indeed the hot path from the code is looking for nearest neighbors. As it has been the case for most systems out there as well

I'll review the code and test this intensively, with all the datasets we know were it's working and come back to you. It might take some time only

PS: thanks for fixing the typos 😂

File: cpp/kiss_icp/core/VoxelHashMap.cpp [hunk 1]
```diff
@@ -49,64 +49,54 @@ namespace kiss_icp {
...
-        Vector3dVector neighboors;
-        neighboors.reserve(27 * max_points_per_voxel_);
-        std::for_each(voxels.cbegin(), voxels.cend(), [&](const auto &voxel) {
-            auto search = map_.find(voxel);
-            if (search != map_.end()) {
-                const auto &points = search->second.points;
-                if (!points.empty()) {
-                    for (const auto &point : points) {
-                        neighboors.emplace_back(point);
+                    auto search = map_.find({i, j, k});
```
Comment (nachovizzo) [inline cpp/kiss_icp/core/VoxelHashMap.cpp:63]: Would you mind being a bit more explicit (on the PR description, not the code) about the change you just made?

I kind of reverse engineered the new solution and I think is better, but if would be nice for the other reviewers to facilitate a bit the high level idea ;)

**Review by nachovizzo — COMMENTED** [review]
Looks good to me! I tried on a few datasets and it's working, although I do not see any real improvement in terms of speed.

I think I prefer the other implementation since the 3 tasks that the lambda function is doing where defined in 3 different steps. With this implementation they all tree tasks got merged into quite a few nested loops, which complicates redability for new users

I will let Tiziano and Benedikt comment and test on this particular PR

Great job @jtaveau

File: cpp/kiss_icp/core/VoxelHashMap.cpp (hunk 1 above)
Comment (jtaveau) [reply cpp/kiss_icp/core/VoxelHashMap.cpp:63]: I agree the readability is not great, I'll profile again the code this evening to see if there is some real improvements and give explanations

**Comment by nachovizzo** [comment]
Closing this to avoid stalling, feel free to re-open when you manage to benchmark it!

Best!

**Pull request closed without merging**